### DISCUSSION TITLE
Add more channel adapters

### DISCUSSION THREAD

**Discussion by inputinput77777** [spring-projects/spring-integration#3720] [Ideas]
I'd like the project to have more channel adapters and be more flexible out of the box. I'm currently using the Reactive mode, but I think these channel adapters should fit both modes:

- Apache Pulsar
- Apache Solr
- Apache Lucene
- Elasticsearch
- Debezium
- HBase
- Cassandra
- Ignite
- Protobuf (!) / Grpc support in general

**Comment by artembilan** [comment]
Hi @inputinput77777 !

Thank you for interest in the project!

We are fully aware that there are not enough channel adapters in Spring Integration.
Only the problem that we are short-handed and just don't have resources and time to investigate and implement all the possible use-cases in the world.

The contribution is welcome though: https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc

Probably it is better to start with them in our extensions project: https://github.com/spring-projects/spring-integration-extensions.

You can find there some initial support for Apache Cassandra though: https://github.com/spring-projects/spring-integration-extensions/tree/main/spring-integration-cassandra.

The GRPC discussion you can find here: https://github.com/spring-projects/spring-integration/issues/2626

Some Elasticsearch is implemented as Spring Cloud Stream Sink application: https://github.com/spring-cloud/stream-applications/tree/main/applications/sink/elasticsearch-sink.
A Debezium support is represented there as a Source: https://github.com/spring-cloud/stream-applications/tree/main/applications/source/cdc-debezium-source.

Probably there are some community efforts in the Internet for others as well, e.g. I see this Spring Boot support for Apache Pulsar: https://github.com/majusko/pulsar-java-spring-boot-starter.

So, in short we are open for supporting more protocols and systems only the problem when our team would be able to dedicate the time for their implementations.
Therefore my ask about contribution.

Thank you again!

**Comment by artembilan** [comment]
The Elasticsearch is discussed here: https://github.com/spring-projects/spring-integration/issues/3527

**Comment by almogtavor** [comment]
Seems pretty messy and hard to follow when there are so many half solutions like that. I think there's should be an issue that will track the creation & status of 10-20 new channel adapters. I also wonder if the repos @artembilan have mentioned are relevant. Would you add to SI a channel adapter that is based on a private community contributor's repo?

**Reply by artembilan** [reply]
Well, at least there is some effort we can benefit from. Those half solutions and those  private community contributor's repos. Everything is good to have and be able to learn from.
Feel free to raise issue for those protocols we don't have issues yet.

I pointed already to Elasticsearch and GRPC.
Apache Cassandra is supported in the extensions project, so anything relevant to it better to raise over there instead.

**Reply by almogtavor** [reply]
I agree that it's good to have, but I think every good channel adapter good implementation that is not in the repo, should move to it/ to the extension one, otherwise it hard to know which channel adapters exists and which should be implemented from zero. Some questions:
1. About the issue, would you prefer an issue that suggests and tracks 10-20 channel adapters, or an 10-20 issues for each one?
2. You'd like the new additions to be in the main repo/ extensions repo? When do you consider a channel adapter ready and agree to move it from the extensions repo to the main?
3. I still don't understand if you are ready to create channel adapters which depends on a community contributor's repo, or would you prefer for example to have the Pulsar connector in the main Spring Integration repo, until Spring will implement the proper infrastructure.

**Reply by artembilan** [reply]
I would say there is no gold rule.
The Spring Integration main project has that `core` module which has everything  to implement some other dedicated module.
The fact that main project has so many modules is just for our, Spring team convenience to maintain them as a single release train.
But this model doesn't stop anyone to have their own independent project which just depend on `spring-integration-core`. For example GCP channel adapters are hosted in the Google project: https://github.com/GoogleCloudPlatform/spring-cloud-gcp.
Microsoft Azure channel adapters are in their own repo: https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/spring.

We have a separate project for AWS channel adapters: https://github.com/spring-projects/spring-integration-aws/.

I believe there might be many more projects in the world which have implementation of channel adapters for some specific protocol.

Others, less popular or like experiments, are in the extensions project I've mentioned before. 
They may move from there to core project or not. It depends on our support or upstream dependency release cycle.
Sometime we do opposite, when the module goes away from Spring support, e.g. that happened in the past with Twitter.

I don't think it is legit to move community repo to ours without agreement with that project commiters. 

No, I would prefer do not have Pulsar (or similar) originally in the main repo. The Extensions one is better to do that experimentation.

So, to summarize: better to raise individual issues per channel adapter for specific protocol. If we don't have one already, like for Elasticsearch and GRPC.
And better to raise them in the extensions project: https://github.com/spring-projects/spring-integration-extensions.
Although we are flexible and can move issues between projects over here on GitHub.
The issue could be like "Add channel adapters for Apache Pulsar". And so on for any other interested protocols and systems.

**Reply by almogtavor** [reply]
@artembilan wouldn't it be much easier and cleaner to work only on one repo? Do you prefer that users would look up for channel adapter in 3 official Spring repos (core, extensions, streaming application) + other specific sources (cloud providers), instead of one? I think the way Apache Camel manages their big repo is much easier to understand and use, to maintain and to contribute to. Maybe it would work better if there would at least be some mapping guide to all of these projects from the documentation. But until I saw this thread, I didn't know that there's a `spring-integration-aws` repo.


I don't quite understand what you suggest about Pulsar. We won't use the community's repo, so we will need some classes that wraps the client in the Spring way (autoconfigurations, beans and more), if you prefer that to be in the extensions repo, does it have any chance to move to the main repo, or would you wait until Spring will have official support to it in all of its relevant projects (like with Mongo, Redis), and only then move to the core repo?

**Reply by artembilan** [reply]
Agreed. It would be easier, but we don't have so many contributors to maintain all the modules in a single repo.
I don't think users look up into repos. They just do Maven search and take those artifacts into their dependency management. 
Plus this is not all: we cannot handle auto-configuration in Spring Integration since it does not depend on Spring Boot, but exactly opposite. 
Spring Integration is just a library proposing building blocks: the rest is up to target project how to use them with or without auto-configuration.
That's why we see many those community-driven and cloud providers based channel adapters. Just because no one stops you to implement them and host yourself.

Yes, I will consider to add some note into the doc about extensions and other projects based on Spring Integration. Or just onto project page: https://spring.io/projects/spring-integration

Since I'm only one over here and I don't know what is Apache Pulsar, I'd prefer to have it in the extensions for now. 
It indeed may move eventually into a main project when it matures enough. And this one already independently if something like Spring for Apache Pulsar will born in the future or not.
For example we had in the past Spring Integration for Apache Kafka project: https://github.com/spring-projects/spring-integration-kafka.
And originally all the supporting components were in this project. Then, after some significant move in Apache Kafka and with some resources on our side we decide to have a Spring for Apache Kafka project.
That Spring Integration were reworked on top of this project at some point. And eventually, when everything has settled in both projects, we have moved it into Spring Integration main project.
I won't mind if something like happens with channel adapters extension for Apache Pulsar.
But it may also stay forever in the extensions project like it happened with Hazelcast and Apache Cassandra channel adapters. Or very popular Zip module is still over there without any questions from community.
Another one taking attention is SMB extension.

Honestly you are the first raising concern that extensions project causes some burden.
It does for us, commiters, when we suffer with releasing at the same time. So, that also becomes a good signal for us to reconsider home for the module.
And one more point for extensions project: you as end-user contribution is just concentrated on a single module you are interested in. 
Those extensions are technically independent projects.
When you do your module in a single repo like main Spring Integration, you need to keep in mind specific for all other modules and their build may annoy you since you are interested only in one module.
This way we find an extensions project is much easier to use for community contributors since they don't have burden of the rest of Spring Integration in their domain.

**Reply by almogtavor** [reply]
I agree with all the things you said accept the last one. Big projects is a standard thing and the splitter project doesn't really help new contributors, since everybody should expect that the build and the clone may take some time when talking about a major project. And as we both agreed, it can be very confusing for a contributor who wants for example to contribute to a specific component. Until I knew of the `extensions` repo, I personally thought SI has only like 10 connectors, while when I entered Camel I saw 300. This misleading thing can make contributors recoil unjustly.

In my opinion if you thing the `zip` or the `cassandra` modules are ready, I would merge them into SI. The current situation gives the effect of an incubating components, and I doubt how many developers who saw that they don't appear in the main repo and in the docs would use this in their production environment.

**Reply by artembilan** [reply]
Sure! We can have a discussion with the team.
But again: the support and releasing boundaries bring their effect how we do things over here.
That is not correct to judge only by the number of modules in the the project.
See the table in the docs: https://docs.spring.io/spring-integration/docs/current/reference/html/endpoint-summary.html#spring-integration-endpoints.
And those only from the official main project.

From now let's stay with our current rules: if you want to contribute new module, start from the extensions repo.

**Reply by almogtavor** [reply]
I personally think it better for the components to be in the same repo, at least for the veteran ones that exists for some months and no bugs found. But what you prefer is what you prefer. Why do you think it's wrong to judge an integration repo based on the number of components? This absolutely is a parameter (of course there are many more, but a user won't care how great is the Redis component of he doesn't use Redis). That's why I personally think it should get more focus.

**Reply by artembilan** [reply]
One module in Spring Integration may have many channel adapters. For example `mail` module has support for all the mail protocols. The "component" in our case is some class(es) correspond(s) to some specific functionality. It does not justify to extract them into a separate module (jar). If you pay closer attention to `core` module, you'll find many components which implements these or those IEP. but still: they are in this single module (jar) - not in the individual managable artifact.

**Reply by almogtavor** [reply]
You right and that is really great. Also the jdbc support. I talk more about the NoSQL databases and other data products (such as Pulsar etc) that doesn't have support, or that have a hidden support (that can be found only after more digging - such as for Elasticsearch/ Cassandra and more)

**Reply by artembilan** [reply]
Sure! That's why I said that I will bring this to the team attention. You gave me a lot of food for thinking.
I'll come back to with some decision later on. 
But now `extensions` is the good place to start contributing.

Thank you again!

**Comment by artembilan** [comment]
Another similar request: https://github.com/spring-projects/spring-integration-extensions/issues/254

**Comment by dinizssa** [comment]
Hi, may I ask you if any plans to add Apache Pulsar adaptor in Spring Integration DSL? I see this discussion is quite old, I'm wondering if I any progress about this matter. Any updates will be appreciated.

**Reply by artembilan** [reply]
There is a dedicated issue for that : https://github.com/spring-projects/spring-integration/issues/3771.
So far we are small team over here and we have some other priorities to deal with.
As usual: contribution is welcome!

**Reply by dinizssa** [reply]
Thanks! Is there any documentation available about creating new incoming/outgoing adapters? Or any adapter we could use as reference? Time is also limited here, but we could check possibilities since we're very interested in this particular topic.

**Reply by artembilan** [reply]
We don’t have such a doc because question about it was never raised . We may think about adding one , probably together with you while you are contributing this feature.

I think Spring Integration Kafka or AMQP modules are closer thing to Pulsar to take inspiration from.

the latest module we developed is gRPC but that one comes with gateways which gave nothing to do with messaging destinations.

thank you!
